File: ruby/OpenRecommender.rb
#!C:\Ruby192\bin\ruby.exe
## WIN32
# --------------------- 
## UNIX
#!/usr/bin/ruby

require 'net/http'
require 'rexml/document'
require 'json'

class OpenRecommender
  
  def initialize
    
  end

  def openrecommenderJSON(url)  
    
    # get the JSON data as a string
    resp = Net::HTTP.get_response(URI.parse(url))
    data = resp.body
    # convert the returned JSON data to native Ruby data structure - a hash
    json = JSON.parse(data)

     # if the hash has 'Error' as a key, we raise an error
    if json.has_key? 'Error'
       raise "HTTP or web service error"
    end

    recommend = "<ul>"
    json['recommendations']['recommendation'].each { |recommendation|
      recommend = recommend + "<li><a href=\"#{recommendation['link']['data']}\" title=\"#{recommendation['description']['data']}\"><img src=\"#{recommendation['image']['data']}\" width=\"120\" height=\"90\"/>#{recommendation['title']['data']}</a></li>\n"
    }
    recommend = recommend + "</ul>"
   
    return recommend
  end


  def openrecommenderXML(url)  

    # get the XML data as a string
    xml_data = Net::HTTP.get_response(URI.parse(url)).body

    # extract event information
    doc = REXML::Document.new(xml_data)
    titles = []
    images = []
    links = []
    descriptions = []
    doc.elements.each('recommendations/recommendation/title') do |ele|
       titles << ele.text
    end
    doc.elements.each('recommendations/recommendation/image') do |ele|
       images << ele.text
    end
    doc.elements.each('recommendations/recommendation/link') do |ele|
       links << ele.text
    end
    doc.elements.each('recommendations/recommendation/description') do |ele|
       descriptions << ele.text
    end

    recommend = "<ul>"
    # print all events
    titles.each_with_index do |title, idx|
       recommend = recommend + "<li><a href=\"#{links[idx]}\" title=\"#{descriptions[idx]}\"><img src=\"#{images[idx]}\" width=\"120\" height=\"90\"/>#{title}</a></li>\n"
    end
    recommend = recommend + "</ul>"

    return recommend
  end

end

File: python/OpenRecommender.py
#!c:/Python27/python.exe -u
#!/usr/bin/env python
# -*- coding: UTF-8 -*-
import json
from xml.dom import minidom
from urllib import urlopen


OPENRECOMMENDER_URL = 'http://openrecommender.org/schema/XML/recommendations.xml' #OpenRecommender Web Service endpoint
OPENRECOMMENDER_NS = 'http://openrecommender.org/schema/XML' #OpenRecommender Namespace

'''
' openrecommenderXML
'   displays XML recommendation data as human-readable HTML   
'''
def openrecommenderXML(u):

  if u is None:
    print("** ERROR - Missing URL (using default) **")
    url = OPENRECOMMENDER_URL
  elif len(u) == 0:
    print("** ERROR - Missing URL (using default) **")
    url = OPENRECOMMENDER_URL
  else:
    url = u
    
  dom = minidom.parse(urlopen(url))
  
  openrecommender = []
  for recommendation in dom.getElementsByTagName('recommendation'): # FOR NAMESPACES, USE: dom.getElementsByTagNameNS(OPENRECOMMENDER_NS, 'recommendation'):
    openrecommender.append(
      '<li><a id="recommendation_'+getRecommendationIdentifier(recommendation)+
      '" href="'+getRecommendationLink(recommendation)+
      '" title="'+getRecommendationDescription(recommendation)+
      '"><img src="'+getRecommendationImage(recommendation)+      
      '"/>'+getRecommendationTitle(recommendation)+'</a></li>\n'
    )    
      
  return '<ul>'+''.join(openrecommender)+'</ul>'


'''
' openrecommenderJSON
'   displays JSON recommendation data as human-readable HTML
'''  
def openrecommenderJSON(u):

  if u is None:
    print("** ERROR - Missing URL (using default) **")
    url = OPENRECOMMENDER_URL
  elif len(u) == 0:
    print("** ERROR - Missing URL (using default) **")
    url = OPENRECOMMENDER_URL
  else:
    url = u
  
  json_string = open(url)
  json_object = json.load(json_string)
    
  openrecommender = []
  for recommendation in json_object["recommendations"]["recommendation"]:    
    id = recommendation["identifier"]
    title = recommendation["title"]["data"]
    image = recommendation["image"]["data"]
    link = recommendation["link"]["data"]
    description = recommendation["description"]["data"]
    
    openrecommender.append(
      '<li><a id="recommendation_'+id+
      '" href="'+link+
      '" title="'+description+
      '"><img src="'+image+
      '" width="120" height="90"/>'+title+'</a></li>\n'
    )
  
  return '<ul>'+''.join(openrecommender)+'</ul>'

  

  
'''
' XML PARSER
'''
#Elements  
def getRecommendationTitle(recommendation):
  return recommendation.getElementsByTagName('title')[0].firstChild.data

def getRecommendationImage(recommendation):
  return recommendation.getElementsByTagName('image')[0].firstChild.data

def getRecommendationLink(recommendation):
  return recommendation.getElementsByTagName('link')[0].firstChild.data
  
def getRecommendationDescription(recommendation):
  return recommendation.getElementsByTagName('description')[0].firstChild.data

  
#<RECOMMENDATION> Attributes
def getRecommendationIdentifier(recommendation):
  return recommendation.getAttribute('identifier')
  
def getRecommendationSender(recommendation):
  return recommendation.getAttribute('sender')  
  
def getRecommendationReceiver(recommendation):
  return recommendation.getAttribute('receivier')
  
def getRecommendationDate(recommendation):
  return recommendation.getAttribute('date')

def getRecommendationTime(recommendation):
  return recommendation.getAttribute('time')
  
def getRecommendationPriority(recommendation):
  return recommendation.getAttribute('priority')
  
def getRecommendationType(recommendation):
  return recommendation.getAttribute('type')
  
  
#<TITLE> Attributes
def getRecommendationTitleDuration(title):
  return title.getAttribute('duration')
  
def getRecommendationTitleLanguage(title):
  return title.getAttribute('language')
  
def getRecommendationTitleCaptions(title):
  return title.getAttribute('captions')
  
def getRecommendationTitleSubtitles(title):
  return title.getAttribute('subtitles')
  
def getRecommendationTitlePlaylist(title):
  return title.getAttribute('playlist')
  
  
#<IMAGE> Attributes
def getRecommendationImageCreator(image):
  return image.getAttribute('creator')
  
def getRecommendationImagePublisher(image):
  return image.getAttribute('publisher')
  
def getRecommendationImageFormat(image):
  return image.getAttribute('format')
  
def getRecommendationImageRights(image):
  return image.getAttribute('rights')
  
def getRecommendationImageRightsHolder(image):
  return image.getAttribute('rightsHolder')


#<LINK> Attributes
def getRecommendationLinkShortlink(link):
  return link.getAttribute('shortlink')
  
def getRecommendationLinkLocation(link):
  return link.getAttribute('location')
  
def getRecommendationLinkMobile(link):
  return link.getAttribute('mobile')
  
def getRecommendationLinkMobiledownload(link):
  return link.getAttribute('mobiledownload')
  
def getRecommendationLinkMobilestream(link):
  return link.getAttribute('mobilestream')

def getRecommendationLinkCreator(link):
  return link.getAttribute('creator')
  
def getRecommendationLinkPublisher(link):
  return link.getAttribute('publisher')
  
def getRecommendationLinkFormat(link):
  return link.getAttribute('format')
  
def getRecommendationLinkRights(link):
  return link.getAttribute('rights')
  
def getRecommendationLinkRightsHolder(link):
  return link.getAttribute('rightsHolder')
  

#<DESCRIPTION> Attributes
def getRecommendationDescriptionRating(description):
  return description.getAttribute('rating')
  
def getRecommendationDescriptionReview(description):
  return description.getAttribute('review')
  
def getRecommendationDescriptionTags(description):
  return description.getAttribute('tags')  